### PR TITLE
raftstore:Fix append log entry oom

### LINKED ISSUE #11379: tikv oom when run sysbench prepare while query under stress
### What version of TiKV are you using?
sh-4.2# ./tikv-server -V
TiKV 
Release Version:   5.3.0-alpha
Edition:           Community
Git Commit Hash:   500ead07966474d280f652dafc4fc4856b99d9e2
Git Commit Branch: heads/refs/tags/v5.3.0
UTC Build Time:    2021-11-09 11:32:30
Rust Version:      rustc 1.56.0-nightly (2faabf579 2021-07-27)
Enable Features:   jemalloc mem-profiling portable sse protobuf-codec test-engines-rocksdb cloud-aws cloud-gcp
Profile:           dist_release

### What operating system and CPU are you using?
8core 16G memory

### Steps to reproduce
TiKVWorkloadStress003
run run sysbench prepare and select *

### What did you expect?
all instance are normal

### What did happened?
one tikv oom several times
image

image

### LINKED ISSUE #11598: Limit append log flow when memory is insufficient

### LINKED ISSUE #12107: Tikv oom after continue run oltp insert for about  4 hours
### What version of TiKV are you using?

[2022/03/07 12:28:00.817 +08:00] [INFO] [client.go:338] ["Cluster version information"] [type=pd] [version=6.0.0-nightly] [git_hash=9bbd6df243b483ac998fb41448d6d71c7dc42eb8]
[2022/03/07 12:28:00.817 +08:00] [INFO] [client.go:338] ["Cluster version information"] [type=tikv] [version=6.0.0-alpha] [git_hash=005efd56d3405b72a26bf623b5895cf3d9da0a6f]
[2022/03/07 12:28:00.817 +08:00] [INFO] [client.go:338] ["Cluster version information"] [type=tidb] [version=6.0.0-nightly] [git_hash=7cf784cf265a1268439a0d401084ffff96b74738]

### Steps to reproduce

run oltp_stability_005, just run oltp insert util disk is full.

### What did you expect?
Everything works fine, no panic and no oom in all components.

### What did happened?
Tikv oom .
image

### PR DESCRIPTION
**Pull request by tier-cap**
### What is changed and how it works?

Issue Number: Close #11598
Close #11379
Close #12107
What's Changed:

```commit-message
when memory usage is high water, add flow control on the append log entries.
disable log entry cache evict functions.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
- Breaking backward compatibility

### Release note 

```release-note
support append log flow control when memory usage high water to avoid oom.
```

### CODE REVIEW

**Comment by 5kbpers** [comment]
Could you open an issue and illustrate the problem this PR solved and the effect of this PR?

File: src/server/service/kv.rs [hunk 1]
```diff
@@ -2106,7 +2108,17 @@ fn needs_reject_raft_append(reject_messages_on_memory_ratio: f64) -> bool {
     let mut usage = 0;
     if memory_usage_reaches_high_water(&mut usage) {
         let raft_msg_usage = (MEMTRACE_RAFT_ENTRIES.sum() + MEMTRACE_RAFT_MESSAGES.sum()) as u64;
-        if raft_msg_usage as f64 > usage as f64 * reject_messages_on_memory_ratio {
+        let cached_entries = get_cached_entries_size();
+        let applying_entries = MEMTRACE_APPLYS.sum() as u64;
+        if (raft_msg_usage + cached_entries + applying_entries) as f64
+            > usage as f64 * reject_messages_on_memory_ratio
+        {
+            info!("need reject log append on memory limit";
```
Comment (innerr) [inline src/server/service/kv.rs:2116]: Maybe `warn` is better?

File: components/raftstore/src/store/memory.rs [hunk 2]
```diff
@@ -71,3 +72,7 @@ pub fn needs_evict_entry_cache(evict_cache_on_memory_ratio: f64) -> bool {
     }
     false
 }
+
+pub fn get_cached_entries_size() -> u64 {
+    RAFT_ENTRIES_CACHES_GAUGE.get() as u64
```
Comment (innerr) [inline components/raftstore/src/store/memory.rs:77]: Looks weird...

File: src/server/service/kv.rs (hunk 1 above)
Comment (tier-cap) [reply src/server/service/kv.rs:2116]: done

**Review by innerr — APPROVED** [review]
LGTM

**Comment by tier-cap** [comment]
> Could you open an issue and illustrate the problem this PR solved and the effect of this PR?

Can refer to #11598

File: src/server/service/kv.rs [hunk 3]
```diff
@@ -2106,7 +2107,17 @@ fn needs_reject_raft_append(reject_messages_on_memory_ratio: f64) -> bool {
     let mut usage = 0;
     if memory_usage_reaches_high_water(&mut usage) {
         let raft_msg_usage = (MEMTRACE_RAFT_ENTRIES.sum() + MEMTRACE_RAFT_MESSAGES.sum()) as u64;
-        if raft_msg_usage as f64 > usage as f64 * reject_messages_on_memory_ratio {
+        let cached_entries = RAFT_ENTRIES_CACHES_GAUGE.get() as u64;
+        let applying_entries = MEMTRACE_APPLYS.sum() as u64;
+        if (raft_msg_usage + cached_entries + applying_entries) as f64
+            > usage as f64 * reject_messages_on_memory_ratio
+        {
+            warn!("need reject log append on memory limit";
```
Comment (5kbpers) [inline src/server/service/kv.rs:2115]: Will this log be too verbose when memory usage is high?

**Review by 5kbpers — COMMENTED** [review]
rest LGTM

File: src/server/service/kv.rs (hunk 1 above)
Comment (tier-cap) [reply src/server/service/kv.rs]: As from minghua's view, we'd better change this to debug.

File: src/server/service/kv.rs (hunk 3 above)
Comment (tier-cap) [reply src/server/service/kv.rs:2115]: Changed to debug based on discussion comments

File: src/server/service/kv.rs (hunk 1 above)
Comment (tier-cap) [reply src/server/service/kv.rs]: Changed to debug based on discussion comments with @5kbpers

**Review by 5kbpers — APPROVED** [review]

**Pull request merged**